akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala

```diff
@@ -4,48 +4,138 @@
+
+  override val shape = SourceShape(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[IOResult]) = {
+    val ioResultPromise = Promise[IOResult]()
+
+    val logic = new GraphStageLogic(shape) with OutHandler {
+      handler ⇒
+      val buffer = ByteBuffer.allocate(chunkSize)
+      val maxReadAhead = inheritedAttributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
+      var channel: FileChannel = _
+      var position = 0L
+      var chunkCallback: Try[Int] ⇒ Unit = _
+      var eofEncountered = false
+      var availableChunks: Vector[ByteString] = Vector.empty[ByteString]
+
+      setHandler(out, this)
+
+      override def preStart(): Unit = {
+        try {
+          // this is a bit weird but required to keep existing semantics
+          require(Files.exists(path), s"Path '$path' does not exist")
+          require(Files.isRegularFile(path), s"Path '$path' is not a regular file")
+          require(Files.isReadable(path), s"Missing read permission for '$path'")
+
+          channel = FileChannel.open(path, StandardOpenOption.READ)
+        } catch {
+          case ex: Exception ⇒
+            ioResultPromise.trySuccess(IOResult(position, Failure(ex)))
+            throw ex
+        }
+      }
+
+      override def onPull(): Unit = {
+        if (availableChunks.size < maxReadAhead && !eofEncountered)
+          availableChunks = readAhead(maxReadAhead, availableChunks)
+        //if already read something and try
+        if (availableChunks.nonEmpty) {
+          emitMultiple(out, availableChunks.iterator,
+            () ⇒ if (eofEncountered) success() else setHandler(out, handler)
+          )
+          availableChunks = Vector.empty[ByteString]
+        } else if (eofEncountered) success()
+      }
+
+      private def success(): Unit = {
+        completeStage()
+        ioResultPromise.trySuccess(IOResult(position, Success(Done)))
+      }
+
+      /** BLOCKING I/O READ */
+      @tailrec def readAhead(maxChunks: Int, chunks: Vector[ByteString]): Vector[ByteString] =
+        if (chunks.size < maxChunks && !eofEncountered) {
+          val readBytes = try channel.read(buffer, position) catch {
+            case NonFatal(ex) ⇒
+              failStage(ex)
+              ioResultPromise.trySuccess(IOResult(position, Failure(ex)))
+              throw ex
+          }
+
+          if (readBytes > 0) {
+            buffer.flip()
+            position += readBytes
+            val newChunks = chunks :+ ByteString.fromByteBuffer(buffer)
+            buffer.clear()
+
+            if (readBytes < chunkSize) {
+              eofEncountered = true
+              newChunks
+            } else readAhead(maxChunks, newChunks)
+          } else {
+            eofEncountered = true
+            chunks
+          }
+        } else chunks
+
+      override def postStop(): Unit = if ((channel ne null) && channel.isOpen) channel.close()
```

Comment:
shouldn't `ioResultPromise` be completed with something here also, in case of abrupt termination